pallets/manta-pay/src/lib.rs

```
@@ -487,42 +506,57 @@ pub mod pallet {
+            max_update: u64,
             receivers: &mut ReceiverChunk,
+            receiver_pulled: &mut u64,
         ) -> bool {
-            let mut iter = if receiver_index == 0 {
-                Shards::<T>::iter_prefix(shard_index)
-            } else {
-                let raw_key = Shards::<T>::hashed_key_for(shard_index, receiver_index as u64 - 1);
-                Shards::<T>::iter_prefix_from(shard_index, raw_key)
-            };
-            for _ in 0..Self::PULL_MAX_PER_SHARD_UPDATE_SIZE {
-                match iter.next() {
-                    Some((_, next)) => receivers.push(next),
+            let max_receiver_index = (receiver_index as u64) + max_update;
+            for idx in (receiver_index as u64)..max_receiver_index {
+                if *receiver_pulled == max_update {
+                    return Shards::<T>::contains_key(shard_index, idx);
+                }
+                match Shards::<T>::try_get(shard_index, idx) {
+                    Ok(next) => {
+                        *receiver_pulled += 1;
+                        receivers.push(next);
+                    }
                     _ => return false,
                 }
             }
-            iter.next().is_some()
+            Shards::<T>::contains_key(shard_index, max_receiver_index)
         }
 
         /// Pulls sender data from the ledger starting at the `sender_index`.
         #[inline]
-        fn pull_senders(sender_index: usize) -> (bool, SenderChunk) {
+        fn pull_senders(sender_index: usize, max_update_request: u64) -> (bool, SenderChunk) {
             let mut senders = Vec::new();
-            let mut iter = VoidNumberSetInsertionOrder::<T>::iter().skip(sender_index);
-            for _ in 0..Self::PULL_MAX_SENDER_UPDATE_SIZE {
-                match iter.next() {
-                    Some((_, next)) => senders.push(next),
+            let max_sender_index = if max_update_request > Self::PULL_MAX_SENDER_UPDATE_SIZE {
```

I think we should either
- just fail the request (and return the node's max update size in the error) instead of silently returning fewer results than the caller expected
or 
- clearly document in the ledger_pull rpc doccomment that this call can return fewer than the requested amount.

---

The name clearly said it is `max_update_request`, it is a "max" value and it is a "request".  Also, please take a look at how RPC works, the returned result contains a boolean field to indicate whether further pull is needed. So failing the request is a bad idea. 

And yes, we need a proper RPC documentation.

---

Failing malformed requests sounds very reasonable to me in Server/Client communication.
Wasn't the main point of this comment tho, I'm happy as long as there's an APIdoc showing correct usage of the call.